### PR TITLE
Retire.js implementation

### PR DESCRIPTION
**Pull request by Sunbus**
Task - https://oat-sa.atlassian.net/browse/TAO-7773 .

Useful links: 
[Retire family](https://github.com/RetireJS)
[our grunt-retire fork](https://github.com/oat-sa/grunt-retire)
[our retire.js fork](https://github.com/oat-sa/retire.js)

It's PR about Retire.js (grunt-retire) implementation for detect the use of JS-library versions with known vulnerabilities.

For this moment grunt-retire have issues with node.js v10+, that's why we are using our forks now. Forks will be switched to the original repo after fix-update from maintainer for our [issue](https://github.com/RetireJS/grunt-retire/issues/40).

How to use: you can get available list of grunt task by use console command `grunt --help` in folder `/tao/views/build`. All retire tasks is marked by description. (Don't forget: you should have installed `node.js` and `npm` for installing, command for installing dependencies - `npm i` )

**UPDATE**
grunt-retire package was updated, our dependency was updated.

### CODE REVIEW

**Comment by krampstudio** [comment]
You can create "Draft pull requests" instead of `WIP`, this will prevent merges (https://github.blog/2019-02-14-introducing-draft-pull-requests)

File: views/build/Gruntfile.js [hunk 1]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
     // load all grunt tasks matching the `grunt-*` pattern
     require('load-grunt-tasks')(grunt);
 
+    const retireInputFiles = extensionHelper.getExtensions().map(item => `${root}/${item}/**/*.js`);
+
+    grunt.initConfig({
+      retire: {
+          js: [retireInputFiles], /** Which js-files to scan. **/
```
Comment (krampstudio) [inline views/build/Gruntfile.js]: Can we do it per globally and per extension ? (like the tasks for the tests and linting)

File: views/build/Gruntfile.js [hunk 2]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
...
+          options: {
+              outputFile: './retire-output.json',
+              jsRepository: 'https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json',
+              nodeRepository: 'https://raw.github.com/RetireJS/retire.js/master/repository/npmrepository.json',
```
Comment (krampstudio) [inline views/build/Gruntfile.js]: It would be great also to figure out the best output format (json or text ?) or being sure it is configurable

File: views/build/Gruntfile.js [hunk 3]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
     // load all grunt tasks matching the `grunt-*` pattern
     require('load-grunt-tasks')(grunt);
 
+    const retireInputFiles = extensionHelper.getExtensions().map(item => `${root}/${item}/**/*.js`);
```
Comment (krampstudio) [inline views/build/Gruntfile.js]: Can we ignore the files in `loader` (the bundles) ?

File: views/build/Gruntfile.js [hunk 4]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
...
+
+    grunt.initConfig({
```
[the remark this answers is not in the document]
Comment (Sunbus) [reply views/build/Gruntfile.js]: done

File: views/build/Gruntfile.js [hunk 5]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
...
+      retire: {
+          js: [retireInputFiles], /** Which js-files to scan. **/
+          options: {
+              outputFile: './retire-output.json',
```
[the remark this answers is not in the document]
Comment (Sunbus) [reply views/build/Gruntfile.js]: done

File: views/build/Gruntfile.js [hunk 6]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
...
+              jsRepository: 'https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json',
```
[the remark this answers is not in the document]
Comment (Sunbus) [reply views/build/Gruntfile.js]: you are right, I forgot to remove it, done

File: views/build/Gruntfile.js (hunk 1 above)
Comment (Sunbus) [reply views/build/Gruntfile.js]: done

File: views/build/Gruntfile.js (hunk 3 above)
Comment (Sunbus) [reply views/build/Gruntfile.js]: it's not clear for me, what exactly I should ignore?

File: views/build/Gruntfile.js [hunk 7]
```diff
@@ -22,6 +22,7 @@ module.exports = function(grunt) {
     var testTasks   = [];
 
     grunt.option('root', root);
+    grunt.option('extensionHelper', extensionHelper);
```
Comment (krampstudio) [inline views/build/Gruntfile.js]: Why did you add it as an option ?

File: views/build/grunt/retire.js [hunk 8]
```diff
@@ -0,0 +1,203 @@
+/* jshint -W055 */
```
Comment (krampstudio) [inline views/build/grunt/retire.js]: Why don't you load this file it using npm ?

File: views/build/grunt/security.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2018 (original work) Open Assessment Technlogies SA
```
Comment (krampstudio) [inline views/build/grunt/security.js]: wrong date in the header

File: views/build/grunt/security.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ */
+
+/**
+ * Main bundle configuration, as well as for the TAO extension
```
Comment (krampstudio) [inline views/build/grunt/security.js]: Wrong comment

File: views/build/grunt/security.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @author Aliaksandr Katovich <aliaksandr@taotesting.com>
+ */
+module.exports = function(grunt) {
+    'use strict';
+    const root = grunt.option('root');
+    const extensionHelper = grunt.option('extensionHelper'); // return all ext names
+
+    grunt.loadNpmTasks('grunt-retire');
```
Comment (krampstudio) [inline views/build/grunt/security.js]: this is not needed grunt tasks are supposed to load by themselves

File: views/build/grunt/security.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+    extensionHelper.getExtensions().forEach(extName => {
```
Comment (krampstudio) [inline views/build/grunt/security.js]: what about 
```js
const root   = grunt.option('root');
const currentExtension = grunt.option('currentExtension');
const reportOutput = grunt.option('reports') || 'reports'; 

grunt.config.merge({
        retire : {
            extension : {
                     js: `${root}/${currentExtension}/**/*.js`
                    options: {
                        outputFile: `./${reportOutput}/${extName}-retire.json`,
                    }
            },
            all : {
                     js: `**/*.js`
                    options: {
                        outputFile: `./${reportOutput}/all-retire.json`,
                    }
            }
        }
});
```

so the style is consistent with the lint task by calling : 
```sh
npx grunt retire:extension --extension=taoQtiItem
```

File: views/build/Gruntfile.js [hunk 13]
```diff
@@ -37,7 +37,21 @@ module.exports = function(grunt) {
...
+          options: {
+              outputFile: './retire-output.json',
+              jsRepository: 'https://raw.github.com/RetireJS/retire.js/master/repository/jsrepository.json',
+              nodeRepository: 'https://raw.github.com/RetireJS/retire.js/master/repository/npmrepository.json',
+          }
+      }
+    });
+
     grunt.loadNpmTasks('@oat-sa/grunt-tao-bundle');
+    grunt.loadNpmTasks('grunt-retire');
```
[the remark this answers is not in the document]
Comment (Sunbus) [reply views/build/Gruntfile.js]: you are right, fixed

File: views/build/Gruntfile.js (hunk 7 above)
Comment (Sunbus) [reply views/build/Gruntfile.js]: hm, missed, that it's our plugin, then fixed

File: views/build/grunt/retire.js (hunk 8 above)
Comment (Sunbus) [reply views/build/grunt/retire.js]: It was experiment, in summary this file was not used, fixed

File: views/build/grunt/security.js (hunk 9 above)
Comment (Sunbus) [reply views/build/grunt/security.js]: fixed

File: views/build/grunt/security.js (hunk 10 above)
Comment (Sunbus) [reply views/build/grunt/security.js]: fixed

File: views/build/grunt/security.js (hunk 11 above)
Comment (Sunbus) [reply views/build/grunt/security.js]: sure, you are right, fixed

File: views/build/grunt/security.js (hunk 12 above)
Comment (Sunbus) [reply views/build/grunt/security.js]: I think that 'npx grunt retiregeneris' is best way, that 'npx grunt retire:extension --extension=generis' because it's shorter, easy to get it from 'grunt --help', in additional it's similar with sass-tasks and etc.

But it's up to you, of course, what should I do?

File: views/build/grunt/security.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2014-2019 (original work) Open Assessment Technlogies SA
```
Comment (krampstudio) [inline views/build/grunt/security.js]: This is a new file, you should set the following header : 
```
* Copyright (c) 2019 (original work) Open Assessment Technlogies SA
```
The range `2014-2019` means the file has been created on 2014 and last modified in 2019

File: views/build/grunt/security.js (hunk 12 above)
Comment (krampstudio) [reply views/build/grunt/security.js]: yes please, this would fit our current model, like for eslint. I'd like to keep the `${extensionName}task` for tasks created with extensions (bundle, sass) and the --extension parameter for global tasks

File: views/build/grunt/security.js (hunk 14 above)
Comment (Sunbus) [reply views/build/grunt/security.js]: Now clear for me, thx! Fixed.
